### PR TITLE
Feature: TypeScript Updates

### LINKED ISSUE #23: v2.0.0 - Typings issue for Breakpoint and useCurrentWidth/useCurrentBreakpointName
The new TypeScript typings break custom breakpoint names for Breakpoint (i.e. sm, md, lg, etc). I currently get the following error:

```
Type '{ children: Element; sm: true; down: true; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Breakpoint> & Readonly<Props> & Readonly<{ children?: ReactNode; }>'.
  Property 'sm' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Breakpoint> & Readonly<Props> & Readonly<{ children?: ReactNode; }>'.
```

Just a suggestion, but either have a breakpoint prop that accepts a string with the following typing (also probably more maintainable):

```
interface Props {
  children?: React.ReactNode;
  breakpoint?: string;
  up?: boolean;
  down?: boolean;
  only?: boolean;
  tagName?: string;
  className?: string;
}
```

Or change the Props typing to the following (which unfortunately lets more in but doesn't restrict the use of the library):

```
interface Props<T extends any> {
  children?: React.ReactNode;
  up?: boolean;
  down?: boolean;
  only?: boolean;
  tagName?: string;
  className?: string;
}
```

Also, you're missing typings for useCurrentWidth and useCurrentBreakpointName:

```
export function useCurrentWidth(): number;
export function userCurrentBreakpointName(): string;
```

### PR DESCRIPTION
**Pull request by dbryantm**
Closes #23 
- Updated typing for Props.
- Added typing for useCurrentWidth().
- Added typing for useCurrentBreakpointName().
- Cleaned up inconsistent indentation within the file.

### CODE REVIEW

File: @types/index.d.ts [hunk 1]
```diff
@@ -4,38 +4,40 @@ export = ReactSocks;
...
+    tagName: string;
+    className: string;
+  }
+
+  export interface BreakpointContextProps {
+    currentWidth: number;
+    currentBreakpointName: string;
+  }
+
+  export function setDefaultBreakpoints(breakpoints: any[]): void;
```
Comment (flexdinesh) [inline @types/index.d.ts]: @dbryantm should this be `breakpoints: string[]`?

I know it was already there but do you think it should be updated as well?

**Comment by flexdinesh** [comment]
@nutboltu Looks good to me. Do you may be want to review this PR?

File: @types/index.d.ts (hunk 1 above)
Comment (dbryantm) [reply @types/index.d.ts]: Yeah, I thought about updating it. Here's an easy solution for that:

```
export function setDefaultBreakpoints(breakpoints: { [key: string]: number }[]): void;
```

It enforces the value to be a number versus a number *or* string... unless I'm reading the README incorrectly and that should allowed? I can definitely drop that in also.

Comment (dbryantm) [reply @types/index.d.ts]: Alright, I updated the typing for setDefaultBreakpoints(). I tested it within my project and you should expect the following:

Good - numbers:
```
setDefaultBreakpoints([ { sm: 0 }, { md: 600 }, { lg: 1200 } ]);
```

Bad - strings - will infer that the property value should be a number:
```
setDefaultBreakpoints([ { sm: '0' }, { md: '600' }, { lg: '1200' } ]);
```

File: @types/index.d.ts [hunk 2]
```diff
@@ -4,38 +4,40 @@ export = ReactSocks;
...
+  }
+
+  export interface BreakpointContextProps {
+    currentWidth: number;
+    currentBreakpointName: string;
+  }
+
+  export function setDefaultBreakpoints(breakpoints: { [key: string]: number }[]): void;
+  export function useCurrentWidth(): number;
+  export function userCurrentBreakpointName(): string;
```
Comment (guschnwg) [inline @types/index.d.ts]: Just a typo here `useCurrentBreakpointName`

Comment (dbryantm) [reply @types/index.d.ts]: Thanks for catching that. I'll update it now.

**Comment by dbryantm** [comment]
@nutboltu I'll just go ahead and export an interface called DefaultBreakpoint. It will look like the following:

```
export interface DefaultBreakpoint { 
  [key: string]: number 
}

export function setDefaultBreakpoints(breakpoints: DefaultBreakpoint[]): void;
```

If you name it Breakpoint you'll get some errors from TS. Otherwise, love the library and would be glad to contribute if you guys need any help.

File: @types/index.d.ts (hunk 2 above)
Comment (dbryantm) [reply @types/index.d.ts]: @guschnwg Fixed. Thanks again!

**Comment by dbryantm** [comment]
I updated the usage of that interface to be more inline with your project's coding style. Now it reads as the following:

```
export interface DefaultBreakpoint {
  [key: string]: number;
}

export function setDefaultBreakpoints(breakpoints: ReactSocks.DefaultBreakpoint[]): void;
```

**Comment by nutboltu** [comment]
I think `DefaultBreakpoint` a bit confusing naming as it seems it might be the `default` type. We can call it either `BreakpointInterface` or `BreakpointType`.

@flexdinesh and @dbryantm  What's your thought?

**Comment by flexdinesh** [comment]
@nutboltu I agree, default sounds a bit misleading. @dbryantm thoughts?

We're close to merge. Once we get a fix on this naming, we could merge and release.

**Comment by dbryantm** [comment]
@flexdinesh @nutboltu, if I had to choose one I would go with `BreakpointType`. I can look at that sometime this morning and pass that along. Thanks!

**Comment by dbryantm** [comment]
@flexdinesh @nutboltu Fixed!

**Comment by flexdinesh** [comment]
Changes published in `v2.0.1`. Thanks @dbryantm @nutboltu @guschnwg! 🎉

**Comment by glook** [comment]
try to set up an breakpoint component, but get only this. Whats can be wrong?
image

**Comment by SugarDarius** [comment]
Hi, 

I'm experiencing the same issue as @glook.
When I take a look to the types I see : 
```typescript
interface Props<T = { [key: string]: boolean }> {
    children?: React.ReactNode;
    up?: boolean;
    down?: boolean;
    only?: boolean;
    tagName?: string;
    className?: string;
  }
```

Maybe there is something missing or wrong ? 

I think we can type like (not really good) : 
```typescript
interface Props {
    [ key: string ]: React.ReactNode | boolean | string;
    children?: React.ReactNode;
    up?: boolean;
    down?: boolean;
    only?: boolean;
    tagName?: string;
    className?: string;
  }
```
Or simply can we declare the default breakpoints ? 
```typescript
interface Props<T = { [key: string]: boolean }> {
    children?: React.ReactNode;
    up?: boolean;
    down?: boolean;
    only?: boolean;
    xsmall?: boolean;
    small?: boolean;
    medium?: boolean;
    large?: boolean;
    xlarge?: boolean;
    tagName?: string;
    className?: string;
  }
```

What do you think guys ? 

Thanks a lot,
Aurélien

**Comment by flexdinesh** [comment]
@glook @sugardarius Would you be able to create a sample repo to replicate this error? Also can you double check if you're using the latest version v2.0.1?

I'll get a fix on this tomorrow.

**Comment by dbryantm** [comment]
@glook @SugarDarius, I've already been looking at the fix for that and should be creating a pull request shortly. It will just be pending review from @flexdinesh and @nutboltu. I believe we're going to have to change the typing of `Props` to be the following:

```
interface Props {
  children?: React.ReactNode;
  up?: boolean;
  down?: boolean;
  only?: boolean;
  tagName?: string;
  className?: string;
  [key: string]: React.ReactNode | string | boolean;
}
```

If you need a temporary fix you can create `react-socks.d.ts` in your `src` folder with the following contents: 

```
import * as React from 'react';

export = ReactSocks;
export as namespace ReactSocks;

declare namespace ReactSocks {
  interface Props {
    children?: React.ReactNode;
    up?: boolean;
    down?: boolean;
    only?: boolean;
    tagName?: string;
    className?: string;
    [key: string]: React.ReactNode | string | boolean;
  }

  interface BreakpointProviderProps {
    children: React.ReactNode;
  }

  interface BreakpointAndModifierProps {
    breakpoint?: string;
    modifier?: string;
    tagName: string;
    className: string;
  }

  export interface BreakpointContextProps {
    currentWidth: number;
    currentBreakpointName: string;
  }

  export interface BreakpointType {
    [breakpoint: string]: number;
  }

  export function setDefaultBreakpoints(breakpoints: ReactSocks.BreakpointType[]): void;
  export function useCurrentWidth(): number;
  export function useCurrentBreakpointName(): string;

  export class Breakpoint extends React.Component<ReactSocks.Props> {
    static extractBreakpointAndModifierFromProps(allProps: ReactSocks.Props): ReactSocks.BreakpointAndModifierProps;
  }

  export class BreakpointProvider extends React.Component<ReactSocks.BreakpointProviderProps> {
    static handleResize(): void;
  }
}
```

**Comment by flexdinesh** [comment]
Thanks @dbryantm. If you could create a PR I'll be able to review it in the morning. We'll verify and release a patch tomorrow.

**Comment by dbryantm** [comment]
@flexdinesh Done. You can view it at #26.

**Comment by flexdinesh** [comment]
@glook @SugarDarius Fix released `2.0.2`.

@dbryantm Thanks for the quick fix on this. Do you think we could create a TypeScript example code in `examples` dir? I would be able to review and run the changes in my local as well before release in case future issues like this pops up.

**Comment by nutboltu** [comment]
@flexdinesh , It would be nice to update our README.md where we need to update the import. as below
```
import { Breakpoint, BreakpointProvider } from 'react-socks';
```

**Comment by SugarDarius** [comment]
Hello Guys, 

Thanks a lot for fixing the issue :D

Have a good day, 
Aurélien

**Pull request merged**